src/System.Buffers.Experimental/System/Buffers/BufferExtensions.cs

```diff
@@ -311,14 +419,291 @@ private static Vector<byte> GetVector(byte vectorByte)
 #endif
         }
 
-        private const ulong xorPowerOfTwoToHighByte = (0x07ul |
-                                                       0x06ul << 8 |
+        private const ulong xorPowerOfTwoToHighByte = (0x07ul       |
+                                                       0x06ul << 8  |
                                                        0x05ul << 16 |
                                                        0x04ul << 24 |
                                                        0x03ul << 32 |
                                                        0x02ul << 40 |
-                                                       0x01ul << 48) + 1;
+                                                       0x01ul << 48 ) + 1;
+        private const ulong flagsToHighByte =         (0x01ul       |
+                                                       0x01ul << 7  |
+                                                       0x01ul << 14 |
+                                                       0x01ul << 21 |
+                                                       0x01ul << 28 |
+                                                       0x01ul << 35 |
+                                                       0x01ul << 42 |
+                                                       0x01ul << 49 );
         private const ulong byteBroadcastToUlong = ~0UL / byte.MaxValue;
         private const ulong filterByteHighBitsInUlong = (byteBroadcastToUlong >> 1) | (byteBroadcastToUlong << (sizeof(ulong) * 8 - 1));
+
+        public struct MatchesEnumerable
+        {
+            private readonly ReadOnlySpan<byte> _buffer;
+            private readonly byte _value;
+            private readonly int _start;
+            private readonly int _upperBound;
+
+            internal MatchesEnumerable(ReadOnlySpan<byte> buffer, byte value)
+            {
+                _buffer = buffer;
+                _value = value;
+                _start = 0;
+                _upperBound = buffer.Length;
+            }
+
+            internal MatchesEnumerable(ReadOnlySpan<byte> buffer, byte value, int start)
+            {
+                _buffer = buffer;
+                _value = value;
+                _start = start;
+                _upperBound = buffer.Length;
+            }
+
+            internal MatchesEnumerable(ReadOnlySpan<byte> buffer, byte value, int start, int length)
+            {
+                _buffer = buffer;
+                _value = value;
+                _start = start;
+                _upperBound = length + start;
+            }
+
+            public MatchesEnumerator GetEnumerator()
+            {
+                return new MatchesEnumerator(_buffer, _value, _start, _upperBound);
+            }
+        }
+
+        public struct MatchesEnumerator
+        {
+            private ReadOnlySpan<byte> _buffer; // don't make readonly, methods called on it
```

Comment:
Meant not to qualify it as a readonly field e.g.

private **readonly** ReadOnlySpan<byte> _buffer;

As the function `.DangerousGetPinnableReference()` is called on it and readonly structs get copied when you call methods on them; so as not to potentially mutate the field struct [which results in a perf hit](https://codeblog.jonskeet.uk/2014/07/16/micro-optimization-the-surprising-inefficiency-of-readonly-fields/)